### ISSUE TITLE
Starting container fails with 'System error: read parent: connection reset by peer'

### ISSUE THREAD

**Issue by meeee** [docker/docker#14203]
On our CI server, we run tests in Docker containers using docker-compose. We link 2-15 containers during one run. We ensure that test jobs running concurrently have different docker-compose project names.

Since we upgraded to Docker 1.7.0 (from 1.6), docker-compose to 1.3.1 (from 1.2), and started killing containers instead of stopping them (faster, remove them anyway), we twice had containers failing to start with the following message from compose:

    Creating x_db_1...
    Creating x_1...
    Cannot start container 10bbc5af8ec0d3bb39b207a6474ec70a0954bff01ff94389684a8b9f52df6067: [8] System error: read parent: connection reset by peer

`/var/log/docker.log` contains the following:

    time="2015-06-25T10:29:44.322521665+02:00" level=info msg="POST /v1.18/containers/19aec1ddb8a5cd771771f16a1f8929bb58eea2cf7e877425a7812f6c6e5756a2/start" 
    time="2015-06-25T10:29:44.690044235+02:00" level=warning msg="signal: killed" 
    time="2015-06-25T10:29:44.915997839+02:00" level=error msg="Handler for POST /containers/{name:.*}/start returned error: Cannot start container 19aec1ddb8a5cd771771f16a1f8929bb58eea2cf7e877425a7812f6c6e5756a2: [8] System error: read parent: connection reset by peer" 
    time="2015-06-25T10:29:44.916111471+02:00" level=error msg="HTTP Error" err="Cannot start container 19aec1ddb8a5cd771771f16a1f8929bb58eea2cf7e877425a7812f6c6e5756a2: [8] System error: read parent: connection reset by peer" statusCode=500 

The container is created, but doesn't start. Trying to manually start it using `docker start` fails with the same error message. Memory is available and the kernel log doesn't show any message from the OOM killer.

Restarting docker temporarily solves the problem, so I assume this is a problem with docker itself, not with docker-compose.

`docker version`:

    Client version: 1.7.0
    Client API version: 1.19
    Go version (client): go1.4.2
    Git commit (client): 0baf609
    OS/Arch (client): linux/amd64
    Server version: 1.7.0
    Server API version: 1.19
    Go version (server): go1.4.2
    Git commit (server): 0baf609
    OS/Arch (server): linux/amd64

`docker info`:

    Containers: 30
    Images: 451
    Storage Driver: aufs
    Root Dir: /var/lib/docker/aufs
    Backing Filesystem: extfs
    Dirs: 517
    Dirperm1 Supported: true
    Execution Driver: native-0.2
    Logging Driver: json-file
    Kernel Version: 3.16.0-0.bpo.4-amd64
    Operating System: Debian GNU/Linux 7 (wheezy)
    CPUs: 4
    Total Memory: 15.52 GiB
    Name: <hostname>
    ID: HYYT:WNZW:UPU7:VI2O:HUTP:EZVV:2MQ2:WCRJ:3SHJ:LZXF:MVLS:P3XC
    WARNING: No memory limit support
    WARNING: No swap limit support

`uname -a`:

    <hostname> 3.16.0-0.bpo.4-amd64 #1 SMP Debian 3.16.7-ckt9-3~deb8u1~bpo70+1 (2015-04-27) x86_64 GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.): Physical machine

Steps to Reproduce:

1. `docker-compose run ...`, `kill` and `rm` 50-200 (estimated) containers with links using docker-compose
2. At some point, docker fails to start a container.

Actual Results: Starting the container fails.

Expected Results: Container starts.

Additional info:

[Kubernetes seemed to have a similar problem](https://github.com/GoogleCloudPlatform/kubernetes/issues/9822), apparently some sort of race. The issue also links to a few occurences where people had a similar problem (mainly on IRC).

**Comment by cjcullen** [comment]
I've included my repro instructions here: https://github.com/GoogleCloudPlatform/kubernetes/issues/9822#issuecomment-118999203

Unfortunately, I don't now how to get docker into the magic state where this is reproducible.

**Comment by dchen1107** [comment]
Please note that the issue we found with kubernetes is with 1.6.2 docker release.

**Comment by shapiroj** [comment]
We see this exact error in several of our containers.  Our only workaround is to rename the container.  Interested in hearing other workarounds or solutions.

**Comment by njuicsgz** [comment]
Any advice for this bug? I suffered it for a long time with docker v1.6.2 in our production environment.

**Comment by cpuguy83** [comment]
Seems like an error connecting to sqlite, or rather while reading it.

**Comment by jjelev** [comment]
Docker 1.7.1 and Ubuntu 14.04. Updated my environment through apt-get and restarted. Nginx container no longer failed to start.

**Comment by meeee** [comment]
@jjelev Yep, as stated in the original description, restarting Docker temporarily fixes the problem. Unfortunately, the problem reappears later.

**Comment by airhorns** [comment]
We're seeing this too, will try and dig up some more information. We're simultaneously executing a lot of the same container, might have something to do with it? Most executions work fine but this has just started happening.

**Comment by airhorns** [comment]
Wow, as has been seen by @cjcullen , the length of the docker run command seems to have something to do with this. I changed the length of one of the -e vars and my deterministic container launch failure went away.

**Comment by adeslade** [comment]
I've been having the same issue. Added an extra space to the failing command and it worked. So odd.

**Comment by rflynn** [comment]
We've hit this bug twice this morning after never seeing it in months of Docker use. Re: string length, we have made tweaks to our command recently.

**Comment by meeee** [comment]
I didn't expect the workaround to work, but after adding a few spaces in an environment variable, we now had a full week without the error occurring. Previously, it occurred almost daily.

We added this environment variable to all our compose containers:
```
DOCKER_FIX: '                                        '
```

**Comment by andrecp** [comment]
I am having this same problem and I same docker version as OP

**Comment by andrecp** [comment]
I also had to add
```environment:
        - DOCKER_FIX='                                        '```
to all my dockerbuild files for this error to go away...

**Comment by chrisjhoughton** [comment]
+1 one for the weird `DOCKER_FIX`!

**Comment by airhorns** [comment]
@burke or @sirupsen do you guys have any ideas on this one?

**Comment by burke** [comment]
Nope, this is weird, I can't imagine what it would be.

**Comment by AlbertodelaCruz** [comment]
+1 adding a FOO env variable. Docker version 1.8.2 and docker-compose 1.4.0.
Curiously, not all hosts suffer this behaviour and need the variable.

**Comment by haoyangz** [comment]
@meeee @andrecp @AlbertodelaCruz @chrisjhoughton Just to clarify, by adding a FOO env variable do you mean adding a line like the following to the Dockerfile?
```
ENV DOCKER_FIX randomvalue
```

I tried this but it doesn't work for me...

**Comment by AlbertodelaCruz** [comment]
Yes, in the docker-compose yaml just to add a foo variable. It seems
problem comes with some kind of command length it results from the yml.

**Comment by LK4D4** [comment]
@mrunalp @crosbymichael any ideas here? Looks superweird, error is definitely from libcontainer.

**Comment by sgirones** [comment]
I've been able to strace the process before being killed. Hope it helps:
```
stat("/bin/sh", {st_dev=makedev(252, 3), st_ino=393238, st_mode=S_IFREG|0755, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=216, st_size=109768, st_atime=2015/10/05-08:55:36, st_mtime=2012/03/29-11:53:23, st_ctime=2015/09/03-15:02:56}) = 0
execve("/bin/sh", ["/bin/sh", "-c", "apt-key add /tmp/key.gpg && rm /tmp/key.gpg"], ["PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", "HOSTNAME=e36c4bc50560", "DEPS_LIST_VERSIONED=autotools-dev:all=20120210.1ubuntu1 bsdmainutils:amd64=8.2.3ubuntu1 ca-certificates:all=20141019ubuntu0.12.04.1 ca-certificates-java:all=20110912ubuntu6 chkconfig:all=11.0-79.1-2 cio-fs-core:amd64=0.0.1-6214 coho-pips:amd64=0.0.1-6199 collectd:amd64=5.4.0-1cohodata4 collectd-core:amd64=5.4.0-1cohodata4 curl:amd64=7.22.0-3ubuntu4.14 debhelper:all=9.20120115ubuntu3 default-jre-headless:amd64=1.6-43ubuntu2 dh-apparmor:all=2.7.102-0ubuntu3.10 ethtool:amd64=3.1-1 flask-resources:all=0.0.5-6025 fontconfig:amd64=2.8.0-3ubuntu9.1 fontconfig-config:all=2.8.0-3ubuntu9.1 fuse:amd64=2.8.6-3cohodata1 gettext:amd64=0.18.1.1-5ubuntu3 gettext-base:amd64=0.18.1.1-5ubuntu3 groff-base:amd64=1.21-7 html2text:amd64=1.3.2a-15 intltool-debian:all=0.35.0+20060710.1 java-common:all=0.43ubuntu2 language-pack-en:all=12.04+20140106 language-pack-en-base:all=12.04+20140106 libaio1:amd64=0.3.109-2ubuntu1 libamd2.2.0:amd64=3.4.0-2ubuntu3 libasn1-8-heimdal:amd64=1.6~git20120311.dfsg.1-2ubuntu0.1 libavahi-client3:amd64=0.6.30-5ubuntu2.1 libavahi-common-data:amd64=0.6.30-5ubuntu2.1 libavahi-common3:amd64=0.6.30-5ubuntu2.1 libblas3gf:amd64=1.2.20110419-2ubuntu1 libboost-dev:amd64=1.48.0.2 libboost-filesystem1.46.1:amd64=1.46.1-7ubuntu3 libboost-program-options1.46.1:amd64=1.46.1-7ubuntu3 libboost-system1.46.1:amd64=1.46.1-7ubuntu3 libboost-thread1.46.1:amd64=1.46.1-7ubuntu3 libboost1.46-dev:amd64=1.46.1-7ubuntu3 libcairo2:amd64=1.10.2-6.1ubuntu3 libcap2:amd64=2.22-1ubuntu3 libcommons-cli-java:all=1.2-3 libcommons-lang-java:all=2.6-3ubuntu1 libcroco3:amd64=0.6.5-1ubuntu0.1 libcups2:amd64=1.5.3-0ubuntu8.7 libcurl3:amd64=7.22.0-3ubuntu4.14 libdatrie1:amd64=0.2.5-3 libdbi1:amd64=0.8.4-5.1 libevent-2.0-5:amd64=2.0.16-stable-1ubuntu0.1 libexpat1-dev:amd64=2.0.1-7.2ubuntu1.2 libfontconfig1:amd64=2.8.0-3ubuntu9.1 libfreetype6:amd64=2.4.8-1ubuntu2.2 libfuse2:amd64=2.8.6-3cohodata1 libgcrypt11:amd64=1.5.0-3ubuntu0.4 libgettextpo0:amd64=0.18.1.1-5ubuntu3 libgfortran3:amd64=4.6.3-1ubuntu5 libgnutls26:amd64=2.12.14-5ubuntu3.9 libgpg-error0:amd64=1.10-2ubuntu1 libgssapi-krb5-2:amd64=1.10+dfsg~beta1-2ubuntu0.6 libgssapi3-heimdal:amd64=1.6~git20120311.dfsg.1-2ubuntu0.1 libgssglue1:amd64=0.3-4ubuntu0.1 libhcrypto4-heimdal:amd64=1.6~git20120311.dfsg.1-2ubuntu0.1 libheimbase1-heimdal:amd64=1.6~git20120311.dfsg.1-2ubuntu0.1 libheimntlm0-heimdal:amd64=1.6~git20120311.dfsg.1-2ubuntu0.1 libhiredis-dev:amd64=0.10.1-2 libhiredis0.10:amd64=0.10.1-2 libhx509-5-heimdal:amd64=1.6~git20120311.dfsg.1-2ubuntu0.1 libidn11:amd64=1.23-2 libjline-java:all=1.0-1 libjpeg-turbo8:amd64=1.1.90+svn733-0ubuntu4.4 libjpeg8:amd64=8c-2ubuntu7 libjs-jquery:all=1.7.1-1ubuntu1 libjson0:amd64=0.9-1ubuntu1.1 libjsoncpp0:amd64=0.6.0~rc2-2 libk5crypto3:amd64=1.10+dfsg~beta1-2ubuntu0.6 libkeyutils1:amd64=1.5.2-2 libkrb5-26-heimdal:amd64=1.6~git20120311.dfsg.1-2ubuntu0.1 libkrb5-3:amd64=1.10+dfsg~beta1-2ubuntu0.6 libkrb5support0:amd64=1.10+dfsg~beta1-2ubuntu0.6 liblapack3gf:amd64=3.3.1-1 liblcms2-2:amd64=2.2+git20110628-2ubuntu3.1 libldap-2.4-2:amd64=2.4.28-1.1ubuntu4.5 libldap2-dev:amd64=2.4.28-1.1ubuntu4.5 liblog4j1.2-java:all=1.2.16-3ubuntu1 libltdl7:amd64=2.4.2-1ubuntu1 liblua5.1-0:amd64=5.1.4-12ubuntu1.1 liblz4-1:amd64=0.0~r129-99cohodata1 libmsgpackc2:amd64=0.5.7-3ubuntu1 libnetty-java:all=3.2.6.Final-2 libnfsidmap2:amd64=0.25-1ubuntu2 libnspr4:amd64=4.10.7-0ubuntu0.12.04.1 libnss3:amd64=3.19.2-0ubuntu0.12.04.1 libnss3-1d:amd64=3.19.2-0ubuntu0.12.04.1 libp11-kit0:amd64=0.12-2ubuntu1 libpango1.0-0:amd64=1.30.0-0ubuntu3.1 libpcap0.8:amd64=1.1.1-10 libpci3:amd64=3.1.8-2ubuntu6 libpcrecpp0:amd64=8.12-4ubuntu0.1 libpcsclite1:amd64=1.7.4-2ubuntu2 libpipeline1:amd64=1.2.1-1 libpixman-1-0:amd64=0.30.2-1ubuntu0.0.0.0.2 libprotobuf7:amd64=2.4.1-1ubuntu2 libpython2.7:amd64=2.7.3-1ubuntu3.8cohodata1 libreadline5:amd64=5.2-11 libroken18-heimdal:amd64=1.6~git20120311.dfsg.1-2ubuntu0.1 librrd4:amd64=1.4.7-2cohodata1 librtmp0:amd64=2.4~20110711.gitc28f1bab-1 libsasl2-2:amd64=2.1.25.dfsg1-3ubuntu0.1 libsasl2-dev:amd64=2.1.25.dfsg1-3ubuntu0.1 libsasl2-modules:amd64=2.1.25.dfsg1-3ubuntu0.1 libsasl2-modules-gssapi-mit:amd64=2.1.25.dfsg1-3ubuntu0.1 libsgutils2-2:amd64=1.33-1 libslf4j-java:all=1.6.4-1 libssl-dev:amd64=1.0.1-4ubuntu5.31 libtasn1-3:amd64=2.10-1ubuntu1.4 libthai-data:all=0.1.16-3 libthai0:amd64=0.1.16-3 libtirpc1:amd64=0.2.2-5 libumfpack5.4.0:amd64=3.4.0-2ubuntu3 libunistring0:amd64=0.9.3-5 libunwind7:amd64=0.99-0.3ubuntu1 liburcu1:amd64=0.8.4-1cohodata7 libwind0-heimdal:amd64=1.6~git20120311.dfsg.1-2ubuntu0.1 libwrap0:amd64=7.6.q-21 libx11-6:amd64=1.4.99.1-0ubuntu2.3 libx11-data:all=1.4.99.1-0ubuntu2.3 libxau6:amd64=1.0.6-4 libxcb-render0:amd64=1.8.1-1ubuntu0.2 libxcb-shm0:amd64=1.8.1-1ubuntu0.2 libxcb1:amd64=1.8.1-1ubuntu0.2 libxdmcp6:amd64=1.1.0-4 libxerces2-java:all=2.11.0-4 libxft2:amd64=2.2.0-3ubuntu2 libxml-commons-external-java:all=1.4.01-2 libxml-commons-resolver1.1-java:all=1.2-7 libxml2:amd64=2.7.8.dfsg-5.1ubuntu4.11 libxrender1:amd64=0.9.6-2ubuntu0.2 libyaml-0-2:amd64=0.1.4-2ubuntu0.12.04.4 libzookeeper-java:all=3.5.0-r1465438+dfsg-60 libzookeeper-mt2:amd64=3.5.0-r1465438+dfsg-60 man-db:amd64=2.6.1-2ubuntu2 mongodb:amd64=2.0.4-1ubuntu2.1 mongodb-clients:amd64=2.0.4-1ubuntu2.1 mongodb-dev:amd64=2.0.4-1ubuntu2.1 mongodb-server:amd64=2.0.4-1ubuntu2.1 netcat-traditional:amd64=1.10-39 nfs-common:amd64=1.2.5-3ubuntu3.2 nmap:amd64=5.21-1.1ubuntu1 openjdk-6-jre-headless:amd64=6b36-1.13.8-0ubuntu1~12.04 openjdk-6-jre-lib:all=6b36-1.13.8-0ubuntu1~12.04 openssl:amd64=1.0.1-4ubuntu5.31 pbzip2:amd64=1.1.6-1 pciutils:amd64=3.1.8-2ubuntu6 po-debconf:all=1.0.16+nmu2ubuntu1 pylint:all=0.25.0-1ubuntu1 python-bson:amd64=2.1-1ubuntu0.1 python-crypto:amd64=2.4.1-1ubuntu0.1 python-daemon:all=1.5.5-1ubuntu1 python-decorator:all=3.3.2-1 python-dev:amd64=2.7.3-0ubuntu2.2 python-dnspython:all=1.9.4-0ubuntu3 python-enum:all=0.4.4-2 python-flask:all=0.8-1 python-httplib2:all=0.7.2-1ubuntu2.1 python-ipaddr:all=2.1.10-1 python-ipy:all=0.75-1 python-jinja2:amd64=2.6-1ubuntu0.1 python-kazoo:all=1.1-cio5 python-keyczar:all=0.6~b.061709+svn502-1 python-ldap:amd64=2.3.13-1build1 python-lockfile:all=0.8-2ubuntu1 python-logilab-astng:all=0.23.0-1ubuntu1 python-logilab-common:all=0.57.1-1ubuntu1 python-markupsafe:amd64=0.15-1 python-mock:all=0.7.2-1 python-netaddr:all=0.7.5-4build2 python-netifaces:amd64=0.6-2ubuntu1 python-nose:all=1.1.2-3 python-novaclient:all=2012.1-0ubuntu1 python-numpy:amd64=1.6.1-6ubuntu1 python-pip:all=1.0-1build1 python-pkg-resources:all=0.6.24-1ubuntu1 python-prettytable:all=0.5-1ubuntu2 python-pyasn1:all=0.0.11a-1ubuntu1 python-pymongo:amd64=2.1-1ubuntu0.1 python-scipy:amd64=0.9.0+dfsg1-1ubuntu2 python-setuptools:all=0.6.24-1ubuntu1 python-support:all=1.0.14ubuntu2 python-tz:all=2011k-0ubuntu5 python-werkzeug:all=0.8.1+dfsg-1 python-yaml:amd64=3.10-2ubuntu0.1 python-zookeeper:amd64=3.5.0-r1465438+dfsg-60 python-zope.interface:amd64=3.6.1-1ubuntu3 python2.7-dev:amd64=2.7.3-1ubuntu3.8cohodata1 redis-py:amd64=2.10.3-1cohodata1 rpcbind:amd64=0.2.0-7ubuntu1.2 rrdcached:amd64=1.4.7-2cohodata1 rrdtool:amd64=1.4.7-2cohodata1 sg3-utils:amd64=1.33-1 socat:amd64=1.7.1.3-1.2 ttf-dejavu-core:all=2.33-2ubuntu1 tzdata-java:all=2015f-0ubuntu0.12.04 zookeeper:all=3.5.0-r1465438+dfsg-60 zookeeperd:all=3.5.0-r1465438+dfsg-60", "DEPS_LIST=autotools-dev bsdmainutils ca-certificates ca-certificates-java chkconfig cio-fs-core coho-pips collectd collectd-core curl debhelper default-jre-headless dh-apparmor ethtool flask-resources fontconfig fontconfig-config fuse gettext gettext-base groff-base html2text intltool-debian java-common language-pack-en language-pack-en-base libaio1 libamd2.2.0 libasn1-8-heimdal libavahi-client3 libavahi-common-data libavahi-common3 libblas3gf libboost-dev libboost-filesystem1.46.1 libboost-program-options1.46.1 libboost-system1.46.1 libboost-thread1.46.1 libboost1.46-dev libcairo2 libcap2 libcommons-cli-java libcommons-lang-java libcroco3 libcups2 libcurl3 libdatrie1 libdbi1 libevent-2.0-5 libexpat1-dev libfontconfig1 libfreetype6 libfuse2 libgcrypt11 libgettextpo0 libgfortran3 libgnutls26 libgpg-error0 libgssapi-krb5-2 libgssapi3-heimdal libgssglue1 libhcrypto4-heimdal libheimbase1-heimdal libheimntlm0-heimdal libhiredis-dev libhiredis0.10 libhx509-5-heimdal libidn11 libjline-java libjpeg-turbo8 libjpeg8 libjs-jquery libjson0 libjsoncpp0 libk5crypto3 libkeyutils1 libkrb5-26-heimdal libkrb5-3 libkrb5support0 liblapack3gf liblcms2-2 libldap-2.4-2 libldap2-dev liblog4j1.2-java libltdl7 liblua5.1-0 liblz4-1 libmsgpackc2 libnetty-java libnfsidmap2 libnspr4 libnss3 libnss3-1d libp11-kit0 libpango1.0-0 libpcap0.8 libpci3 libpcrecpp0 libpcsclite1 libpipeline1 libpixman-1-0 libprotobuf7 libpython2.7 libreadline5 libroken18-heimdal librrd4 librtmp0 libsasl2-2 libsasl2-dev libsasl2-modules libsasl2-modules-gssapi-mit libsgutils2-2 libslf4j-java libssl-dev libtasn1-3 libthai-data libthai0 libtirpc1 libumfpack5.4.0 libunistring0 libunwind7 liburcu1 libwind0-heimdal libwrap0 libx11-6 libx11-data libxau6 libxcb-render0 libxcb-shm0 libxcb1 libxdmcp6 libxerces2-java libxft2 libxml-commons-external-java libxml-commons-resolver1.1-java libxml2 libxrender1 libyaml-0-2 libzookeeper-java libzookeeper-mt2 man-db mongodb mongodb-clients mongodb-dev mongodb-server netcat-traditional nfs-common nmap openjdk-6-jre-headless openjdk-6-jre-lib openssl pbzip2 pciutils po-debconf pylint python-bson python-crypto python-daemon python-decorator python-dev python-dnspython python-enum python-flask python-httplib2 python-ipaddr python-ipy python-jinja2 python-kazoo python-keyczar python-ldap python-lockfile python-logilab-astng python-logilab-common python-markupsafe python-mock python-netaddr python-netifaces python-nose python-novaclient python-numpy python-pip python-pkg-resources python-prettytable python-pyasn1 python-pymongo python-scipy python-setuptools python-support python-tz python-werkzeug python-yaml python-zookeeper python-zope.interface python2.7-dev redis-py rpcbind rrdcached rrdtool sg3-utils socat ttf-dejavu-core tzdata-java zookeeper zookeeperd", "HOME=/root"]) = 0
+++ killed by SIGKILL +++
```

I first thought it might be related to "Limits on size of arguments and environment" (http://man7.org/linux/man-pages/man2/execve.2.html), but I haven't been able to reproduce when creating a container with the same environment (the error above happens during a docker build)

**Comment by mrunalp** [comment]
@LK4D4 This could happen if the parent process dies. Maybe a strace -f with runc on an affected system might help to figure this out.

**Comment by delcypher** [comment]
Just to note I'm experiencing this issue with Docker 1.8.3.

Seems fairly reproducible. I have a python script that uses docker-py and it rapidly creates and then starts very short lived containers in parallel. It seems starting the containers seems to occasionally fail.

If I have time I might try to condense it down to a single python script that others can try.

**Comment by sjfink** [comment]
+1 for DOCKER_FIX.    scary.

**Comment by opskumu** [comment]
+1 for DOCKER_FIX. docker 1.6.2. But why? It's really very strange ... ...

**Comment by delcypher** [comment]
Well I tried writing a script that stress testing docker by starting lots of very short lived containers and then destroying them in parallel but unfortunately I can't reproduce the issue any more (with my script or with my original set up where I first observed the bug).

I did find another issue though ( #17661).

**Comment by ross-w** [comment]
Popping in here to say I've just experienced this issue on 1.8.3.

**Comment by Nalum** [comment]
This just started happening for me. Can confirm adding another environment variable `FOO="    "` got it working again.

```
Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:43:42 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:43:42 UTC 2015
 OS/Arch:      linux/amd64
```

**Comment by theonlydoo** [comment]
Why adding a useless environment variable fixes this issue ?

**Comment by Nalum** [comment]
Happened to me again changing the value by a space fixed it again

**Comment by tobowers** [comment]
We just saw this in a Docker-in-Docker container running 1.8.2 inside a host running 1.8.2 (host is centos 7.1 and container is based on centos:7 image). I added the DOCKER_FIX and it started working again.

Restarting the container running docker (hence restarting docker) did *not* fix the problem.

**Comment by oliviermadre** [comment]
Same case here.

I have a docker-compose.yml file having ~10 services to start.

When I do `docker-compose -p blabla up -d`, one of them is crashing (always the same).
When I do `docker-compose up -d`, it's not crashing (notice that the containers won't have the same prefix)

When I try the DOCKER_FIX weird fix, and then tries `docker-compose -p blabla up -d`, it's not crashing
For me, changing the prefix or adding a dummy environment variable fix the problem.

**Comment by xeor** [comment]
I have the exact same problem, but it hits 1 of my containers that I start with `docker-compose up -d`.
It was 100% reproducible, I did several stop, rm, up on that container. It always failed starting up.

* Stopping the container, doing rm on it, changing the `restart` policy from `always` to `never`, did not fix the problem.
* Stopping the container, changing the restart policy, starting it up again, DID fix the problem. The container started.
* Stopping the same container, deleting it, then tried to start it, brought the same error back.

The trick with using a long environment-variable also solved the trick, but with that fix, I could also delete the container and start it up again without any problems.

Server Version: 1.9.0
Storage Driver: overlay
Backing Filesystem: xfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.3.0-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)

**Comment by thaerlo** [comment]
Same problem here, fixed with that weird workaround too.
Docker version: 1.9.0
Kernel version:  3.10.0-123.20.1.el7.x86_64
OS version:       CentOS Linux release 7.1.1503 (Core)

**Comment by demi-rluddy** [comment]
I'm seeing the same issue, but none of the workarounds mentioned above seem to be helping.

The same container is failing for me on multiple machines, including one with 1.8.1 and one with 1.9.1:

```
3.13.0-48-generic

Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:35:49 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:35:49 UTC 2015
 OS/Arch:      linux/amd64
```
```
3.13.0-68-generic

Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64
```

Of note:
I have several nearly-identical containers (different only in version name) that are working fine. The failing one has the longest name.

Everything in my system has very long names, which normally I would think is irrelevant but with the spooky workaround many people are seeing work may matter:
* My images are registryhost:registryport/company-project-image:build
* My container names are hostname-containername-20characterid

**Comment by theonlydoo** [comment]
Have you tried to reproduce with another kernel ?

**Comment by demi-rluddy** [comment]
In my case, the issue appeared to be related to memory limits. There was a message in dmesg that the task was being killed as a result of a limit, although elsewhere did not report it as such the way it usually does.

I suspect it may have been hitting the memory limit in the middle of creating the container, before it was prepared to report it.

Increasing the memory limit immediately caused the otherwise-identical task to immediately work as expected on both test systems.

Thank you for your patience - I would have investigated sooner but was unavailable during the holiday.

**Comment by salomvary** [comment]
The magic environment variable "fixed" it for us too. :crying_cat_face:

**Comment by scottmuc** [comment]
FYI the following version exhibits this issue:

```
Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 12:59:02 UTC 2015
 OS/Arch:      linux/amd64
```

but we did not see this issue with:

```
Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:20:14 UTC 2015
 OS/Arch:      linux/amd64
```

That being said... I'm not 100% sure because reproducing the issue is sketchy at best. Our worship of the magical white space has fixed our build and we eagerly await a resolution to this issue.

**Comment by denis-yuen** [comment]
FYI, since it wasn't 100% clear to me, this issue affects plain `docker run` commands, not just containers launched with docker compose. 

The "fix" above, adding `ENV DOCKER_FIX='                                        '` to a Dockerfile does indeed seem to work.

**Comment by javicacheiro** [comment]
Same problem here, I have a cluster all running the same image and the same version of docker (1.7.1) and it just affects one node.

Instead of adding the env variable to the Dockerfile and having to rebuild and push the image, I just rerun adding a DOCKER_FIX env var (or other var name also works):
````
docker run -e DOCKER_FIX='' ...
````

**Comment by gesellix** [comment]
Is the foo env variable the only one? Or does anyone already have other env variables configured?

**Comment by theonlydoo** [comment]
the 'fix' just consists to add an empty var, no matter its name

**Comment by gesellix** [comment]
Question is: does any var suffice or does it have to be one with whitespace? Maybe it's only about adding the very first var, which forces some piece of code to run, which would otherwise not run, but could be necessary to start a container.

**Comment by scottmuc** [comment]
@gesellix in our case the problem appeared to be an off-by-one error. It turned out that the number of characters in one env var mattered. When the value was 2 characters, the issue manifested itself. If that value was 1 or 3 characters, things worked. If possible, I'll try and dig up the full docker command and to if there's a pattern with the number of characters the whole command is and how it potentially relates to his issue.

We added DOCKER_FIX to simply have a commentable section of code to explain and link to this specific GH issue.

**Comment by gesellix** [comment]
@scottmuc maybe it would be interesting to see the `docker inspect <container>` output (especially for the `Env` property) after the container has been created. As far as I read this issue, the container doesn't start, but can be created successfully. Comparing the different `docker inspect` output we might find a test case which deterministically breaks.

**Comment by Confusion** [comment]
I remember reading a report about a  bug hunt where the bug turned out to be triggered by the total size of the shell environment variables (of course this is from memory, so exact details may be wrong). This sounds like it is the same thing and would also explain why adding an environment variable matters. Some quick googling didn't turn anything up yet though :/

**Comment by scottmuc** [comment]
@gesellix @Confusion Unfortunately I'm going on holidays, but if this is still open, I'm pretty sure I can set aside a day between Christmas and New Years to attempt to recreate the environment and provide some more meaningful debug output. We were able to reliably recreate the problem.

**Comment by gesellix** [comment]
@scottmuc nice!

Not sure if it's relevant or even comparable: what happens if the same shell/command being executed on container start would be executed during a `docker build` (via `RUN`)? Does that break, too?

**Comment by cjcullen** [comment]
@Confusion I documented my struggles over here: https://github.com/kubernetes/kubernetes/issues/9822. The simplest I got my repro was:

Docker run a container w/ host networking (this step only needs to happen once):
```
# docker run -d --net=host gcr.io/google_containers/busybox /bin/sh -c "sleep 100000000"
5cd36d0a9d7e86101be05e72e0d138e3fde09a194a52b28b2b678223f9c70258
```
Post to docker's containers/create endpoint with a 1289 character Entrypoint (`>>` is actually `\u003e\u003e`):
```
# curl -H "Content-Type: application/json" --unix-socket /var/run/docker.sock -X POST -d '{"Image":"
gcr.io/google_containers/busybox","Entrypoint":["/bin/sh","-c","echo 0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdefghi6abcdefghi7abcdefghi8abcdefghi9abcdefghi0abcdefghi1abcdefghi2abcdefghi3
abcdefghi4abcdefghi5abcdef 1>>/a"]}' http:/containers/create
{"Id":"b71cb237f1d6f219892716f649668be593ecabe8523a7c62a54b7085dc9d9582","Warnings":null}
```
Start the container w/
1) `--net` and `--ipc` pointing to the host-networked container from above.
2) CgroupParent set to "/", MemorySwap set to -1.
3) 13 Binds, 4 characters each.
```
# curl -H "Content-Type: application/json" --unix-socket /var/run/docker.sock -X POST -d '{"Binds":[
"/:/a","/:/b","/:/c","/:/d","/:/e","/:/f","/:/g","/:/h","/:/i","/:/j","/:/k","/:/l","/:/m"],"Network
Mode":"container:5cd3","IpcMode":"container:5cd3","CgroupParent":"/","MemorySwap":-1}' http:/contain
ers/b71cb237f1d6f219892716f649668be593ecabe8523a7c62a54b7085dc9d9582/start
Cannot start container b71cb237f1d6f219892716f649668be593ecabe8523a7c62a54b7085dc9d9582: [8] System 
error: read parent: connection reset by peer
```

For me, the 1341 characters between Entrypoint and Binds (1289 + 13*4), `--net`, `--ipc`, `CgroupParent`, `MemorySwap`, and exact # of Binds were necessary for it to reproduce. Adding a character to a Bind and removing one from the Entrypoint still caused the error. Dropping the number of Binds to 12, I couldn't find the magic number to reproduce it.

**Comment by alexisvincent** [comment]
Status on this?

**Comment by scottmuc** [comment]
@gesellix @Confusion argh, I tried to recreate the issue this morning and I can't seem to reproduce it. I wish I could provide more details. If I catch the bug again in the wild I'll get as much info as I can.

**Comment by w00251574** [comment]
When I use kubelet to create and delete pod for several times, I also occur a similar problem like this.
When the problem  happened, I saw the pod had been in “ContainerCreating” status for a long time. Then I saw the kubelet' s log have the keywords:  "System error: read parent: connection reset by peer"

[root@localhost ~]# kubectl get pod
NAME         READY     STATUS              RESTARTS   NETWORKS             AGE
test-0x8c3   1/1       Running             0          ,network1,subnet1    18m
test-31luj   1/1       Running             0          ,network1,subnet1    18m
test-ecjvi   1/1       Running             0          ,network1,subnet1    18m
test-fhnh6   1/1       Running             0          ,network1,subnet1    18m
test-hisnz   1/1       Running             0          ,network1,subnet1    18m
test-j8cz1   0/1       ContainerCreating   0          ,network1,subnet1    18m
test-krv1e   1/1       Running             0          ,network1,subnet1    18m
test-n1sgo   1/1       Running             0          ,network1,subnet1    18m
test-nanf3   1/1       Running             0          ,network1,subnet1    18m
test-yo0q9   1/1       Running             0          ,network1,subnet1    18m

Log:
[root@localhost ~]# journalctl -u kubelet | grep test-j8cz1
Dec 28 11:14:38  kubelet[55064]: E1228 11:14:38.276002 55064 manager.go:1871] Failed to setup networking for pod "test-j8cz1_default" using network plugins: API error (500): Cannot start container 2f7cde17b74888cf54b7f6e7dabf2534b045fdca67a4708a7970683f79271c48: [8] System error: read parent: connection reset by peer
Dec 28 11:14:38  kubelet[55064]: W1228 11:14:38.277147 55064 manager.go:1874] Clear infra container network failed for pod "test-j8cz1_default": Cannot find endpoint information related to container

**Comment by treyhyde** [comment]
adding a FOO=BAR also worked for me.   Different people on different machines may or may not get this error with the same docker-compose.yml and same versions of the container same versions of docker-machine, docker, etc, even from fresh docker-machine instances.

**Comment by jstangroome** [comment]
This issue has been affecting some of our Production systems for several months. We only use `docker run`, not docker-compose.

On some nodes we run 300+ containers based on only 3 or 4 images. Only occasionally will a particular container be affected by this issue but that container will then be affected on all nodes in the same cluster (all nodes in a cluster run all the same containers).

Moving that container to a different node in a *different* cluster will typically fix this issue.

Restarting the `docker` daemon on the node will fix the issue but is obviously not a great option in Production.

Adding a placeholder environment variable to the affected container's configuration will fix the issue.

**Comment by derjust** [comment]
I'm unsure if that really relates, but for those using AWS ECS and also get this error:
For me it was caused by CPU/Memory set to 0,1 oder 2 (either due to defaulting or explicitly).
Setting a "reasonable" explicit value like 1024 solved the issue for me.

**Comment by burke** [comment]
Apparently this has recently started to burn us in a more serious way one a broader range of hosts after upgrading to 1.9.1.

**Comment by thaJeztah** [comment]
Thanks for letting us know @burke!

**Comment by csfrancis** [comment]
**tl;dr; Data is being left in a pipe - https://github.com/docker/docker/commit/be7ef6370a4bbe1c333182cecf1b1d783d193aba is a workaround.**

I believe I’ve found the root cause of this issue. The bug is not in Docker, but is in the [opencontainers/runc](https://github.com/opencontainers/runc) library that docker depends on. As @burke mentioned, we’ve been seeing this a lot more in production in 1.9.1, and having a reproducible test case made tracking this down a lot easier. In our case the problem persisted across restarts of the docker daemon and even across reboots.

Attaching to the docker daemon with `strace`, I was able to find the syscall that generated the `ECONNRESET`:
```
[pid 30291] execve("/sbin/borginit", ["/sbin/borginit", "resque-worker"], [/* 29 vars */] <unfinished ...>
[pid 30295] +++ exited with 0 +++
[pid 30294] +++ exited with 0 +++
[pid 30293] +++ exited with 0 +++
[pid 30292] +++ exited with 0 +++
[pid 30291] <... execve resumed> )      = 0
[pid 26317] <... read resumed> 0xc20f88b600, 512) = -1 ECONNRESET (Connection reset by peer)

```

Not surprisingly, it's a `read` call. Looking at what other operations were performed on that file descriptor (`18`) gleaned a bit more information:
```
[pid 26317] write(18, "{\"args\":[\"/sbin/borginit\",\"resqu"..., 7681 <unfinished ...>
[pid 30294] <... futex resumed> )       = 0
[pid 26317] <... write resumed> )       = 7681
[pid 30294] futex(0xc208020b58, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid 30291] socket(PF_LOCAL, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0 <unfinished ...>
[pid 26317] shutdown(18, SHUT_WR <unfinished ...>
[pid 30291] <... socket resumed> )      = 4
[pid 26317] <... shutdown resumed> )    = 0
[pid 30291] setsockopt(4, SOL_SOCKET, SO_BROADCAST, [1], 4 <unfinished ...>
[pid 26317] read(18,  <unfinished ...>
```

Also, what was exactly was file descriptor 18? After a bit of searching I was able to find it was being returned from a call to `socketpair`:
```
[pid 26245] socketpair(PF_LOCAL, SOCK_STREAM|SOCK_CLOEXEC, 0, [17, 18]) = 0
```

At this point I had enough information from the `strace` log (what created the fd, how and what order it was used) to start going through the docker source to try to find corresponding source code. After some searching I found [where the `socketpair` syscall](https://github.com/docker/docker/blob/v1.9.1/vendor/src/github.com/opencontainers/runc/libcontainer/process_linux.go#L221-L232) originated, and [where operations are performed on the pipe](https://github.com/docker/docker/blob/v1.9.1/vendor/src/github.com/opencontainers/runc/libcontainer/process_linux.go#L178). The code that generates the error is:
```go
  if err := json.NewDecoder(p.parentPipe).Decode(&ierr); err != nil && err != io.EOF {
    return newSystemError(err)
  }
```

With this information, I now tried to piece together logically how we ended up in this bad state:

1. We create a pipe
2. Write data to that pipe
3. Shutdown the pipe for further writes
4. Wait for a read response

However, step 4 is what fails. How can `read` fail with `ECONNRESET`? According to [this Stackoverflow answer](http://stackoverflow.com/a/2979806), on a local pipe/socket, if the socket is closed on the other side before all written data is read, a subsequent `read` would return `ECONNRESET`. Going with this hypothesis, I tried to find what was reading on the other side of the pipe. To figure that out, I searched for the same bytes that we `write` to the pipe earlier in the `strace` log, which yields:
```
[pid 30291] read(3, "{\"args\":[\"/sbin/borginit\",\"resqu"..., 512) = 512
[pid 30292] <... select resumed> )      = 0 (Timeout)
[pid 30291] read(3,  <unfinished ...>
[pid 30292] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid 30291] <... read resumed> "2.36:8125\",\"STATSD_IMPLEMENTATIO"..., 1024) = 1024
[pid 30291] read(3,  <unfinished ...>
[pid 30292] <... select resumed> )      = 0 (Timeout)
[pid 30291] <... read resumed> "mount_cmds\":null},{\"source\":\"tmp"..., 2048) = 2048
[pid 30292] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid 30291] read(3, "ation_flags\":null,\"data\":\"\",\"rel"..., 4096) = 4096
```

What immediately struck me as somewhat suspicious was the fact that all of the calls to `read` were returning a power of 2. More digging through the code and I found [where this data was consumed](https://github.com/docker/docker/blob/v1.9.1/vendor/src/github.com/opencontainers/runc/libcontainer/factory_linux.go#L212-L243). From what I can gather, when docker starts a new container it creates a pipe where the daemon writes out a JSON configuration blob, which is read by the spawned container.

One thing I had also done is run `strace` on a successful container launch (using the workaround of adding an extra environment variable). The difference was the last call to `read` did not returned `11` bytes and not the entire size of the read buffer:
```
[pid 24238] read(3, "{\"args\":[\"/sbin/borginit\",\"resqu"..., 512) = 512
[pid 24239] <... select resumed> )      = 0 (Timeout)
[pid 24239] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid 24238] read(3, "2.36:8125\",\"STATSD_IMPLEMENTATIO"..., 1024) = 1024
[pid 24239] <... select resumed> )      = 0 (Timeout)
[pid 24238] read(3,  <unfinished ...>
[pid 24239] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid 24238] <... read resumed> "null,\"postmount_cmds\":null},{\"so"..., 2048) = 2048
[pid 24238] read(3,  <unfinished ...>
[pid 24239] <... select resumed> )      = 0 (Timeout)
[pid 24239] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid 24238] <... read resumed> "80,\"propagation_flags\":null,\"dat"..., 4096) = 4096
[pid 24239] <... select resumed> )      = 0 (Timeout)
[pid 24239] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid 24238] read(3, "_count\":0}\n", 8192) = 11
```

All of this led me to believe that there might have been some data sitting the pipe that was not being read before the pipe was closed, which leads to the connection reset error. A [simple hack](https://github.com/docker/docker/commit/be7ef6370a4bbe1c333182cecf1b1d783d193aba) showed that this was in fact the case, AND what was was left in the buffer. After recompiling docker and trying to run a container with `strace` attached to the daemon, we can see exactly what's going on:
```
[pid 27263] read(3,  <unfinished ...>
[pid 27263] <... read resumed> "2.36:8125\",\"STATSD_IMPLEMENTATIO"..., 1024) = 1024
[pid 27263] read(3,  <unfinished ...>
[pid 27263] <... read resumed> "mount_cmds\":null},{\"source\":\"tmp"..., 2048) = 2048
[pid 27263] read(3,  <unfinished ...>
[pid 27263] <... read resumed> "ation_flags\":null,\"data\":\"\",\"rel"..., 4096) = 4096
[...snip...]
[pid 27263] read(3, "\n", 2048)         = 1
```
**BINGO**. There's a newline in the pipe that isn't being read! My assumption as to what's happening:

1. The new container [reads its configuration as JSON](https://github.com/docker/docker/blob/v1.9.1/vendor/src/github.com/opencontainers/runc/libcontainer/init_linux.go#L63)
2. The JSON decoder does not read the final `\n` from the pipe, because as far as it's concerned it has completed unmarshaling.
3. The pipe is closed, which causes the parent to fail with `ECONNRESET` when it tries to read back a response.

What's the best way to fix this? I leave this up to the Docker team. I can think of two options off the top of my head:

- Use my disgusting hack
- Remove the newline from the JSON encoded configuration

**Comment by csfrancis** [comment]
I've created https://github.com/opencontainers/runc/pull/508 to fix this upstream.

**Comment by thaJeztah** [comment]
@csfrancis wow, awesome work! :heart:

**Comment by airhorns** [comment]
@csfrancis way to go!!! Did you figure out what the pattern is that triggers this only sometimes and only for some line lengths? It seems weird that the unmarshaller would have different state that triggered consuming that last token only sometimes.

**Comment by csfrancis** [comment]
@airhorns It has to do with the total size of the JSON configuration that's passed down from the docker daemon to the container. This is influenced by the line length of the command, specifically, the reason why the "add the environment variable" workaround works is because adding that environment variable modifies the total size of the JSON.

The underlying data stream being fed into the JSON decoder is being buffered (as demonstrated by the `read` syscalls). That stream starts at 512 bytes and is doubled every time the decoder needs more data. 99.9% of the time this works correctly because the trailing newline at the end of the JSON is read from the pipe into the buffer. It's that 0.1% of the time where the **last** character in the underlying buffer (the buffer that is being filled by the `read` syscall) is the final `}`. At that point, the JSON decoder says, "I'm done!" and doesn't ask for any more data, even though there's still a newline that was added by the encoder that's sitting in the pipe.

IMO the real fix for this would be for the Golang JSON decoder to try to read the final newline when decoding JSON, since they're the ones that add the newline during encoding.

**Comment by hairyhenderson** [comment]
@csfrancis - whoa, nice sleuthing. Always good to see your mad `strace` skills :grin: (also, hi! :wave:)

**Comment by scottmuc** [comment]
@csfrancis very interesting! Thank you for going into how you troubleshooted this.

**Comment by MMore** [comment]
@csfrancis really good explanation. Awesome work!

**Comment by hqhq** [comment]
@csfrancis I think your analysis is excellent and almost 100% right :+1: , but there might be other easier work around.

Can you test the following patch see if this will fix the problem as well? (Though the runc master branch has changed the synchronisation, we can still test for current docker version)
I think another race condition could cause this error is that parent closed the write pipe and then child tried to read when `err != nil`. And we did consider the dirty pipe condition and tried to clean it, but it's in another path which didn't cover the scenario you described.

```
$ git diff
diff --git a/vendor/src/github.com/opencontainers/runc/libcontainer/factory_linux.go b/vendor/src/github.com/opencontainers/runc/libcontainer/factory_linux.go
index 70513f7..4610e58 100644
--- a/vendor/src/github.com/opencontainers/runc/libcontainer/factory_linux.go
+++ b/vendor/src/github.com/opencontainers/runc/libcontainer/factory_linux.go
@@ -223,12 +223,13 @@ func (l *LinuxFactory) StartInitialization() (err error) {
        // specific env vars.
        os.Clearenv()
        defer func() {
+               // ensure that any data sent from the parent is consumed so it doesn't
+               // receive ECONNRESET when the child writes to the pipe.
+               ioutil.ReadAll(pipe)
+
                // if we have an error during the initialization of the container's init then send it back to the
                // parent process in the form of an initError.
                if err != nil {
-                       // ensure that any data sent from the parent is consumed so it doesn't
-                       // receive ECONNRESET when the child writes to the pipe.
-                       ioutil.ReadAll(pipe)
                        if err := json.NewEncoder(pipe).Encode(newSystemError(err)); err != nil {
                                panic(err)
                        }
diff --git a/vendor/src/github.com/opencontainers/runc/libcontainer/process_linux.go b/vendor/src/github.com/opencontainers/runc/libcontainer/process_linux.go
index 114c71b..d89ad0b 100644
--- a/vendor/src/github.com/opencontainers/runc/libcontainer/process_linux.go
+++ b/vendor/src/github.com/opencontainers/runc/libcontainer/process_linux.go
@@ -273,8 +273,7 @@ func (p *initProcess) sendConfig() error {
        if err := json.NewEncoder(p.parentPipe).Encode(p.config); err != nil {
                return err
        }
-       // shutdown writes for the parent side of the pipe
-       return syscall.Shutdown(int(p.parentPipe.Fd()), syscall.SHUT_WR)
+       return nil
 }

 func (p *initProcess) createNetworkInterfaces() error {
```

**Comment by cyphar** [comment]
@csfrancis This seems like a very weird stdlib bug to me, have you tried opening an issue upstream about it?

**Comment by wader** [comment]
I guess the newline comes from https://github.com/golang/go/blob/f78a4c84ac8ed44aaf331989aa32e40081fd8f13/src/encoding/json/stream.go#L193
@csfrancis Nice work!

**Comment by thaJeztah** [comment]
There's a PR to work around this issue here; https://github.com/opencontainers/runc/pull/515

**Comment by crosbymichael** [comment]
@thaJeztah its not a work around, its a fix!!! how dare you!

**Comment by thaJeztah** [comment]
@crosbymichael ~~worked around~~ fixed my comment

**Comment by cyphar** [comment]
+1 for "major kick-ass fix". I wish I could label all of my PRs like that.

**Comment by extemporalgenome** [comment]
@cyphar the trailing newline that the Go stdlib json streaming encoder produces is intentional (both to make http://ndjson.org/ trivial, and because text output should generally use newlines as a line _terminator_ anyway).

The streaming decoder stopping at the '}' (and not reading the '\n') is also intentional, so that you can embed json in another format, and use the json decoder to read "just far enough but not too far".

**Comment by hqhq** [comment]
@extemporalgenome If all of these are intentional, how to explain the problem in this issue? Is that because of the wrong usage? Then what's the right usage about pipe communication with encoder and decoder? Should users always use an extra readall before they close the pipe?

I understand the intentions, just think it didn't consider all use cases. But I doubt there is perfect solution.

**Comment by cyphar** [comment]
@extemporalgenome The actual behaviour is intentional (the reason cited being that you need to separate numbers so the encoder knows when a number is finished), but the consequences clearly are not thought out. IMO the correct solution for Go's side would be that the "newline appending" is **only** done for JSON encoding of numbers (not objects, strings or lists). The decoder simply *has* to read the next character in that case (numbers don't have a delimiter).

I'd also like to point out that doing something like `json.Decode` inside of an XML decoding from a file is a **bad practice** for many reasons (what if your JSON decoder is badly written and reads too many characters? not to mention the fact that the JSON parser has no context from the XML decoder). The correct way to do it would be to store the JSON as a value in the XML, unmarshal the XML and then unmarshal the JSON.

So, no. I don't agree that the stdlib "does the right thing" here.

**Comment by extemporalgenome** [comment]
There seems to be the notion, both in the submitted fix and in discussions here, that there's "arbitrary junk" in the pipe after the end of the JSON. It's not junk by any means -- it's a single newline (that by the way will always, consistently be there). Fixing it by using json.Marshal instead of json.Encoder is fine (in which case `}{` may appear as a byte sequence in the pipe at some point, which again, is fine). It can also be fixed by doing a single byte read after calling Decode, assuming the top-level json entity on the wire is not a number (and why would it be?). Outstanding work by @csfrancis on diagnosing this; the solution in either case is the very definition of trivial (which is often the case with the hardest-to-find bugs).

Miscellany follows...

---

> IMO the correct solution for Go's side would be that the "newline appending" is only done for JSON encoding of numbers (not objects, strings or lists)

@cyphar having inconsistent *encoding* behavior for a text format is rarely a good idea. From the perspective of the majority use case, the Go streaming encoder currently does the right thing. If you want precise reading without rewind/peeking in the _general_ case, you must either use a size-tagged format (TLV binary or netstrings), or have some unambiguous sentinel value (such as '}' or ']') _without_ buffered reading.

Fully-precise reading is not a stated goal of the Go streaming *decoder* (and as you mentioned is impossible for JSON in general). It will stop as soon as it can based on the parse/scan state (so it will stop precisely on quotes, braces, brackets, null, true, and false, but not numbers). Also, the vast majority of JSON has an object or array as the root entity. Designing against the rarest case (numbers as root entities in a stream) would not be useful. Unread, non-semantic data actually causing an error is certainly a very rare edge case -- most json-using applications have absolutely no opportunity to experience this class of bug.

> The correct way to do it would be to store the JSON as a value in the XML, unmarshal the XML and then unmarshal the JSON.

Yes, absolutely. What I described is more useful for space separated, line-delimited ad-hoc formats, rather than something well-structured like XML, i.e.  `:atom1 {"some":"json"} :atom2` -- in such cases you'd typically use json.Marshal to write, and json.Decoder to read.

> We should probably push for a fix upstream.

Likely not going to happen. json.NewEncoder did not originally append newlines -- they intentionally added that later, after discussion and general consensus. Removing the newlines would break _lots_ of existing applications, both in Go, and non-Go consumers of Go services.

> Your example of ndjson is very odd too.

It's not my example -- it's an extremely well-adopted format that has independently been "discovered" by many people (there's also http://jsonlines.org/ as an effectively identical independent spec). A top-level array of objects is semantically very different than a line-delimited sequence of top-level objects. It's the same reason YAML has streams (http://www.yaml.org/spec/1.2/spec.html#id2801681).

**Comment by rachit1arora** [comment]
hello , i have also encountered this problem in docker version 1.9.1 . 
I am encountering this in our production environment when we do a 
docker exec <conatinerId> <command> and not during the docker run . 

I know that the fix is avilable in docker 1.10 but is there a way to get the fix in docker 1.9.1 ? We may not be able to migrate to docker 1.10 soon . 

Is there a work around we can try in docker exec command ? Many people reported that docker run -e DOCKER_FIX='' worked for them . 
How Can we resolve this in docker exec command ?

**Comment by extemporalgenome** [comment]
You could try passing in a longer command line, for example:

docker exec the-container sh -c "intended-command args # extra padding junk
in comment"

**Comment by utamar** [comment]
hey @crosbymichael , what's the ETA for this issue?

**Comment by loretoparisi** [comment]
I get the same error with this `DockerFile`

```
FROM ubuntu:16.04
COPY . /app
VOLUME /app
```

I was doing a long running task using `tensorflow` with 

`./nvidia-docker-run --volumes-from myImage --rm -it tensorflow/tensorflow:0.10.0-gpu bash`

**Comment by vielmetti** [comment]
I'm getting the error

`System error: json: cannot unmarshal object into Go value of type libcontainer.syncType.`

on CoreOS. docker -v reports `Docker version 1.10.3, build 1f8f545`

**Comment by thaJeztah** [comment]
@vielmetti looks like possibly the JSON of one of your containers got corrupted, might be worth trying to find which one and either remove that container, or try to fix the JSON. Also keep in mind that CoreOS ships with a modified version of Docker (see https://github.com/coreos/docker/commit/1f8f545 for the commit it's built from), and issues should be reported in their issue tracker first

**Comment by vielmetti** [comment]
Thanks @thaJeztah , I opened a CoreOS issue which appears to be unrelated to this particular issue (the error text is the same but the reproduction is different).

**Comment by shawntoxu** [comment]
Which version of this bug can be resolved ？？？？

**Comment by thaJeztah** [comment]
@shawntoxu see the milestone attached to this issue; it's in docker 1.10.0

**Comment by shawntoxu** [comment]
thx. 在2017年06月29日 08:35，Sebastiaan van Stijn 写道： @shawntoxu see the milestone attached to this issue; it's in docker 1.10.0 — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by jfdoerre** [comment]
We are still seeing this sporadic error "... [9] System error: read parent: connection reset by peer" with docker 1.10.3. Maybe it is, because we are on CentOS7.2?

It seems the fix of this error requires that you are using opencontainers/runc, but I guess we don't use that in our setup.
What we are using are the docker rpms from the CentOS repository:
`# cat /etc/centos-release
CentOS Linux release 7.2.1511 (Core) 
# uname -a
Linux ilgnext-jenkins.svl.ibm.com 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
# docker -v
Docker version 1.10.3, build 9419b24-unsupported`

Any chances that this error is fixed in a more recent version for CentOS, e.g. in CentOS7.3?

Or else, is there a solid way to test, if on my upgraded system the error does no longer appear.

**Comment by thaJeztah** [comment]
@jfdoerre docker 1.10 is no longer maintained, and the version in the CentOS repository is the Red Hat fork of docker, for which the code doesn't live in this repository.

> Any chances that this error is fixed in a more recent version for CentOS, e.g. in CentOS7.3?

Be aware that CentOS is a rolling release, which means that when 7.3 was released, 7.2 no longer receives updates, so it's indeed recommended to be on the current version

If you're still seeing this on the current (17.03 or 17.06) release of the official Docker packages, please open a new issue.

**Comment by thaJeztah** [comment]
I'm locking the conversation on this issue, because the original issue was resolved; if you encounter this issue on an up to date version of docker, please open a new issue instead.
